ndpyramid/reproject.py

```diff
@@ -0,0 +1,202 @@
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        The dataset to create a multiscale pyramid of.
+    projection : str, optional
+        The projection to use. Default is 'web-mercator'.
+    level : int
+        The level of the pyramid to create.
+    pixels_per_tile : int, optional
+        Number of pixels per tile
+    resampling : dict
+        Rasterio warp resampling method to use. Keys are variable names and values are warp resampling methods.
+    extra_dim : str, optional
+        The name of the extra dimension to iterate over. Default is None.
+
+    Returns
+    -------
+    xr.Dataset
+        The multiscale pyramid level.
+
+    Warning
+    -------
+    Pyramid generation by level is experimental and subject to change.
+    """
+    save_kwargs = {'pixels_per_tile': pixels_per_tile}
+    attrs = {
+        'multiscales': multiscales_template(
+            datasets=[{'path': '.', 'level': level}],
+            type='reduce',
+            method='pyramid_reproject',
+            version=get_version(),
+            kwargs=save_kwargs,
+        )
+    }
+    dim = 2**level * pixels_per_tile
+    projection_model = Projection(name=projection)
+    dst_transform = projection_model.transform(dim=dim)
+
+    # Convert resampling from string to dictionary if necessary
+    if isinstance(resampling, str):
+        resampling_dict: dict = defaultdict(lambda: resampling)
+    else:
+        resampling_dict = resampling
+
+    # create the data array for each level
+    ds_level = xr.Dataset(attrs=ds.attrs)
+    for k, da in ds.items():
+        if clear_attrs:
+            da.attrs.clear()
+        if len(da.shape) == 4:
+            # if extra_dim is not specified, raise an error
+            if extra_dim is None:
+                raise ValueError("must specify 'extra_dim' to iterate over 4d data")
+            da_all = []
+            for index in ds[extra_dim]:
+                # reproject each index of the 4th dimension
+                da_reprojected = _da_reproject(
+                    da.sel({extra_dim: index}),
+                    dim=dim,
+                    crs=projection_model._crs,
+                    resampling=Resampling[resampling_dict[k]],
+                    transform=dst_transform,
+                )
+                da_all.append(da_reprojected)
+            ds_level[k] = xr.concat(da_all, ds[extra_dim])
+        else:
+            # if the data array is not 4D, just reproject it
+            ds_level[k] = _da_reproject(
+                da,
+                dim=dim,
+                crs=projection_model._crs,
+                resampling=Resampling[resampling_dict[k]],
+                transform=dst_transform,
+            )
+    ds_level.attrs['multiscales'] = attrs['multiscales']
+    return ds_level
+
+
+def pyramid_reproject(
+    ds: xr.Dataset,
+    *,
+    projection: typing.Literal['web-mercator', 'equidistant-cylindrical'] = 'web-mercator',
```

Comment:
Since this is shared between `pyramid_reproject` and `level_reproject` is it worth defining somewhere and sharing?

Reply:
Good call! That will make it easier if we expand the options.